### PR TITLE
Gallery block: Add images size selector

### LINKED ISSUE #1450: Gallery: choosing thumbnail size
This is not a great experience in core at the moment, but since some themes / people may be relying on specific thumbnail generation for galleries, it'd be good to offer that option in the inspector, to choose a specific thumbnail size from the registered ones.

### PR DESCRIPTION
**Pull request by mmtr**
## Description

Add an image size selector to the gallery block that allows users to choose the size of the images. It defaults to the large size.

Fixes #1450

## How has this been tested?
Manually, following the steps below:
- Insert a gallery block and add some images.
- Verify there is a new images size selector that defaults to "Large".
- Choose a different size.
- Verify the editor loads the images in the selected size.
- Publish the post and verify the frontend loads the images in the selected size.
- Open an existing post containing a gallery block and verify images in large size are still loaded.

## Screenshots 

Screen Shot 2019-11-18 at 17 02 40

## Types of changes
Enhancement.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

File: packages/block-library/src/gallery/edit.js [hunk 1]
```diff
@@ -224,6 +229,25 @@ class GalleryEdit extends Component {
...
+	getImagesSizeOptions() {
+		const { imageSizes } = this.props;
+		return map( imageSizes, ( { name, slug } ) => ( { value: slug, label: name } ) );
+	}
+
+	updateImagesSize( sizeSlug ) {
+		const { attributes: { images }, resizedImages } = this.props;
+
+		const updatedImages = map( images, ( image ) => {
+			const url = get( resizedImages, [ parseInt( image.id, 10 ), sizeSlug ] );
```
Comment (youknowriad) [inline packages/block-library/src/gallery/edit.js]: I'm probably not the best person to review this PR but I know that Galleries can have images that are just URLs not coming from the media library. these should be taken into consideration (ignored properly)

File: packages/block-library/src/gallery/edit.js [hunk 2]
```diff
@@ -377,10 +416,41 @@ class GalleryEdit extends Component {
...
+			mediaUpload,
+		} = getSettings();
+
+		const resizedImages = reduce(
+			ids,
+			( currentResizedImages, id ) => {
+				if ( ! id ) {
+					return currentResizedImages;
+				}
+				const image = getMedia( id );
```
Comment (jorgefilipecosta) [inline packages/block-library/src/gallery/edit.js]: This is marking an HTTP request per image, if a post as multiple galleries with multiple images we may end up with issue https://github.com/WordPress/gutenberg/issues/17465. I guess we could apply a similar solution to https://github.com/WordPress/gutenberg/pull/17504.

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @mmtr, thank you for this PR. It worked great on my tests. I just left a comment related to the number of network requests we could be making.

File: packages/block-library/src/gallery/edit.js (hunk 2 above)
Comment (mmtr) [reply packages/block-library/src/gallery/edit.js]: Done in 231da67

**Comment by idea--list** [comment]
Applied these changes on my localhost, seems to pass my manual tests.
This is an awesome PR, fixes a long standing severe issue with GB.
(Also cover and media&text blocks would need this setting)

**Review by jorgefilipecosta — APPROVED** [review]
Worked well on my tests 👍

**Comment by ShareTextures** [comment]
@mmtr I installed Gutenberg 7.2 but I can't see image size selector?

**Comment by mmtr** [comment]
> @mmtr I installed Gutenberg 7.2 but I can't see image size selector?

It only shows up when the gallery block contains an image coming from the media library. If the gallery block is empty or contains only images that are just URLs, the selector won't cause any effect and thus is hidden.

Can you verify you are not using the gallery block under one of these scenarios?

**Comment by ShareTextures** [comment]
> > @mmtr I installed Gutenberg 7.2 but I can't see image size selector?
> 
> It only shows up when the gallery block contains an image coming from the media library. If the gallery block is empty or contains only images that are just URLs, the selector won't cause any effect and thus is hidden.
> 
> Can you verify you are not using the gallery block under one of these scenarios?

It's interesting I tried it another computer and it's show it now. Cookies possible :) Thanks

**Comment by mmtr** [comment]
Let's move the discussion to https://github.com/WordPress/gutenberg/issues/19556

**Pull request merged**